src/layers/io/input/input_layer.cpp

```diff
@@ -42,6 +45,18 @@ input_distconv_adapter(Layer& layer, const bool shuffle_required)
   for (int i = 0; i < layer.get_num_children(); ++i) {
     m_is_input_processed.push_back(layer.get_child_layers()[i]->distconv_enabled());
   }
+  auto &l = dynamic_cast<input_layer<
+    TensorDataType, T_io_buffer, T_layout, Dev>&>(this->layer());
+  // TODO: hdf5_reader is assumed to return a sub-sample partitioned
+  // in the same way as specified by the parallel strategy of this input
+  // layer. Other data readers are assumed to return a complete
+  // sample, thus shuffling is required (unless sample-parallel
+  // strategy is given). Conceptually, it seems to make sense if a
+  // data reader is annotated with a parallel strategy. Note that,
+  // when the HDF5 data reader is used, it is assumed that it is used
+  // in all execution modes.
+  auto training_dr = l.get_data_reader(execution_mode::training);
+  m_shuffle_required = dynamic_cast<hdf5_reader<TensorDataType>*>(training_dr) == nullptr;
```

Comment:
@bvanessen Do we actually need this change? This raises an exception because it tries to get the training execution context, which is set after the setup pass.
I have fixed this issue in 632941e1a754a0445ee9b7265661ca0fbabefcbb (and removed in ed5a8ab98af66e39dafbb620b7bd0b9f3a7ca411) by passing DataReaderMetaData to the adapter which contains the information whether shuffle is required.

Reply:
No I don't think that we need that.  I reverted that deletion because I thought that it was due to a bad merge.  If it shouldn't be there, please remove it again.